### PR TITLE
chore: improve annotations

### PR DESCRIPTION
**Pull request by aarnphm**
As I was working with cattrs recently, I notice there are some places where
annotations are still missing.

Hopefully this PR address most of the popular API to be annotated.

### CODE REVIEW

**Review by Tinche — CHANGES_REQUESTED** [review]
Left a few comments :)

I would also suggest rebasing on top of `tin/copy` instead of `main`.

Thanks!

File: src/cattrs/converters.py [hunk 1]
```diff
@@ -340,19 +340,23 @@ def _unstructure_union(self, obj):
 
     # Python primitives to classes.
 
-    def _structure_error(self, _, cl):
+    def _structure_error(self, _, cl: Any) -> None:
```
[the remark this answers is not in the document]
Comment (aarnphm) [reply src/cattrs/converters.py]: `NoReturn` is available since 3.6

**Comment by aarnphm** [comment]
cc @Tinche whenever you have time. This is just dx improvement, so it would be nice to include in 22.2. Otherwise there is no rush.

**Comment by aarnphm** [comment]
I will close this now since I believe you also update annotations on the copy branch.

**Comment by Tinche** [comment]
Ah sorry, the copy branch took a lot longer than I was expecting. I don't think I have all of these though, if you reopen I will merge in.

**Comment by aarnphm** [comment]
sg, let me check for any merge config.

**Pull request closed without merging**